DNA rework

phil235 opened the pull request:
- Fixed monkey starting with no dna. (I had to move genetics setup in the object subsystem)
- Fixed bugs with monkeyize/humanize: Fixes #7803 (humanized monkey nobloodtype)
- Fixes #9298 , monkeyed ling have troubles humanizing themself
- Fixes despawning clothes when monkeyizing. Fixes #11855
- Replaced check_dna_integrity proc by simpler has_dna proc when required.
- created set_species() proc
- fixed space retrovirus not transfering SE (despite having a domutcheck()).
- I renamed mecha/var/dna to dna_lock to avoid confusion
- I renamed an armor var in a species proc to armor_block to avoid confusion with species/var/armor.
- I removed many if(dna) checks in lots of files.
- I removed duplicate defense procs between human/proc/X and dna.species/proc/X since dna is now always set.
- Anatomic panacea from changeling removes alien embryo correctly. Fixes #6247
- Fixes runtime when trying to put dna-less brain mmi into a dnalocked mech.
- Removed carbon/var/list/features, we now only have dna.features and prefs.features
- <b>Remove hulk mutation from lizards and other species (Fixes #6413); only real humans can acquire hulk. (less work on sprites for each ones, fixes lizard tail not in hulk color)</b>
- Fixes cloning not setting up correctly dna UE and dna.real_name
- I fixed the issue with sucked+cloned ling being unable to absorb
- I fixed issue with changeling proc checking if they have the dna already not working.
- Fixes #4095, low health hulk with DAA getting stuck in loop of acquiring/losing hulk.
- I added a second layer for mutations to differientate mutations that go below and above the body layer (Fixes #7858). Removed the Hulk's red eyes, since they are now hidden behind the mob's eyes.
- Fixes #10048, the transform to initial appearence button was fucking up the dna.
- Fixes changeling readaptation not removing chameleon skin, arm blade, organic suits and shield (Fixes #10567)
- Moved dna scanner and dna console code out of dna.dm and into their own files under the machinery folder.
- Fixed laser eyes mutations not showing its overlay.
- Monkeys now have blood types and are compatible with humans (mostly because having two separate kinds of blood is more difficult to do)
- Fixes #11927
- Fixed lesser form changeling, when humanized via dna injector, the ling no longer keeps its Human Form power.
- Dna now gets qdel'd, just like reagents, when the mob gets qdel'd.
- Probably some other stuff I forgot.

phil235 committed:
- Fixed monkey starting with no dna.
- Fixed bugs with monkeyize/humanize: 7803 (humanized monkey nobloodtype)
- Fixes 9298 monkeyed ling have troubles humanizing themself (already fixed?)
- Fixes despawning clothes when monkeyizing. 11855
- Replaced check_dna_integrity proc by simpler has_dna proc when required.
- created set_species() proc
- fixed space retrovirus not transfering SE (despite having a domutcheck()). Still need to check if it needs a name = real_name.
- I renamed mecha/var/dna to dna_lock to avoid confusion
- I renamed an armor var in a species proc to armor_block to avoid confusion with species/var/armor.
- I removed many if(dna) checks in lots of files.
- I removed duplicate defense procs between human/proc/X and dna.species/proc/X since dna is now always set.
- Anatomic panacea from changeling removes alien embryo correctly. 6247
- Fixes runtime when trying to put dna-less brain mmi into a dnalocked mech.
- Removed carbon/var/list/features, we now only have dna.features and prefs.features
- Remove hulk mutation from lizards and other species (Fixed 6413); only real humans can acquire hulk. (less work on sprites for each ones, fixes lizard tail not in hulk color)
- Fixes cloning not setting up correctly dna UE and dna.real_name
- I fixed the issue with sucked+cloned ling being unable to absorb
- I fixed issue with changeling proc checking if they have the dna already not working.
- Fixed 4095, low health hulk with DAA getting stuck in loop of acquiring/losing hulk.
- I added a second layer for mutations to differientate mutations that go below and above the body layer (Fixes 7858)
- Fixes 10048, the transform to initial appearence button was fucking up the dna.

- Fixes cloning not setting up correctly dna UE and dna.real_name
- Fixed the issue with sucked+cloned ling being unable to absorb
- Fixed issue with changeling proc checking if they have the dna already not working.
- Fixed 4095, low health hulk with DAA getting stuck in loop of acquiring/losing hulk.
- Added a second layer for mutations to differientate mutations that go below and above the body layer (Fixes 7858)
- Fixes 10048, the transform to initial appearence button was fucking up the dna.

phil235 committed:
Tiny mistake.

Core0verload commented:
Is the bug when UI injection wasn't updating race fixed too?

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into DnaMonkeyStuff

Conflicts:
	code/__HELPERS/unsorted.dm
	code/modules/admin/verbs/one_click_antag.dm

WJohn commented:
A GORILLION FIXES

phil235 committed:
Fixes the errors from merge conflicts
Fixes changeling readaptation not removing chameleon skin, arm blade, organic suits and shield.

phil235 commented on code/__DEFINES/misc.dm:
```diff
@@ -49,7 +50,7 @@
 #define FIRE_LAYER				1		//If you're on fire
-#define TOTAL_LAYERS			26		//KEEP THIS UP-TO-DATE OR SHIT WILL BREAK ;_;
+#define TOTAL_LAYERS			27		//KEEP THIS UP-TO-DATE OR SHIT WILL BREAK ;_;
 
```
Having two layers for mutations is required for two reasons: 
- We want to be able to have both hulk and laser eyes shown at the same time. 
- We want hulk to be behind body layer (to fix the underwear bug) but laser eyes must appear above the mob's eyes.

phil235 commented on code/controllers/subsystem/objects.dm:
```diff
@@ -15,6 +15,7 @@ var/datum/subsystem/objects/SSobj
 	NEW_SS_GLOBAL(SSobj)
 
 /datum/subsystem/objects/Initialize(timeofday, zlevel)
+	setupGenetics()
```
The objects subsystem that calls the atom movables' initialize() including monkey/initialize() which make the monkey's dna and requires genetics to already be setup. If genetics setup was left in the ticker subsystem (which is initialized after the SSObj) it would fuck up the monkey dna initialization.

Iamgoofball replied:
why dont you put these in as code comments

phil235 replied:
Most of these are too long to keep in or are about comments I'm removing. I've added plenty of comments in the code to explain some of the new code.

phil235 commented on code/datums/martial.dm:
```diff
@@ -104,24 +92,18 @@
-	playsound(D.loc, 'sound/weapons/punch1.ogg', 25, 1, -1)
-
+	playsound(D.loc, A.dna.species.attack_sound, 25, 1, -1)
 
```
Fixed a lot of copypasta and simplified the code because the dna checks are no longer needed.

phil235 commented on code/game/dna.dm:
```diff
@@ -266,691 +309,20 @@
-	return 1
-/////////////////////////// DNA MISC-PROCS
-
-
```
dna scanner and console (below) are moved in their own files under machinery.

phil235 commented on code/game/gamemodes/changeling/powers/lesserform.dm:
```diff
@@ -12,9 +12,7 @@
 	user << "<span class='warning'>Our genes cry out!</span>"
 
-	user.monkeyize(TR_KEEPITEMS | TR_KEEPIMPLANTS | TR_KEEPORGANS | TR_KEEPDAMAGE | TR_KEEPSE | TR_KEEPSRC)
+	user.monkeyize(TR_KEEPITEMS | TR_KEEPIMPLANTS | TR_KEEPORGANS | TR_KEEPDAMAGE | TR_KEEPVIRUS | TR_KEEPSE)
```
Changeling lesser form / human form doesn't get you rid of viruses anymore (that's anatomic panacea's job).

phil235 commented on code/game/gamemodes/changeling/powers/panacea.dm:
```diff
@@ -16,6 +16,9 @@
+	for(var/obj/item/organ/internal/I in user)
+		if(istype(I,/obj/item/organ/internal/body_egg/alien_embryo))
+			I.Remove(user)
+			I.loc = get_turf(user)
```
Anatomic Panacea gets rid of alien embryos.

phil235 commented on code/game/gamemodes/changeling/powers/tiny_prick.dm:
```diff
@@ -92,15 +92,17 @@
-		C.do_jitter_animation(500)
-		C.take_organ_damage(20, 0) //The process is extremely painful
-
-	target.visible_message("<span class='danger'>[target] begins to violenty convulse!</span>","<span class='userdanger'>You feel a tiny prick and a begin to uncontrollably convulse!</span>")
```
this should only happen when the target is a carbon.

phil235 commented on code/modules/mob/interactive.dm:
```diff
@@ -86,22 +86,14 @@
-	facial_hair_style = random_facial_hair_style(gender)
-	hair_color = random_short_color()
-	facial_hair_color = hair_color
-	eye_color = "blue"
```
all this is already done in the randomize_human() called by human/New()

phil235 commented on code/modules/mob/living/carbon/carbon_defines.dm:
```diff
@@ -9,9 +9,6 @@
 
-	var/list/features = list("mcolor" = "FFF", "tail" = "Smooth", "snout" = "Round", "horns" = "None", "frills" = "None", "spines" = "None", "body_markings" = "None")
-
-
```
we don't need both a carbon/var/features and a datum/dna/var/features.

Incoming5643 replied:
The only reason features is also carbon defined is because we can't call up a character's DNA in the preference screen before the mob exists. If you fixed/can fix that then yeah it's more than welcome to go.

phil235 replied:
There is a separate features var for /datum/preferences, this is the one that changes when you modify you features in the lobby window. By the time preferenecs/copy_to(human) is called the human mob has already been created. So there's no need for a carbon/var/features, I've replcaed all instances of it by dna.features w/o issue.

phil235 commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -16,6 +16,12 @@
+	create_dna(src)
+	randomize_human(src)
+	dna.initialize_dna()
+
```
Moved before organ initilization because some organ Remove/Insert proc call dna stuff.

phil235 commented on code/modules/mob/living/carbon/human/human_damage.dm:
```diff
@@ -160,38 +160,5 @@
-		// Will set our damageoverlay icon to the next level, which will then be set back to the normal level the next mob.Life().
-
-		updatehealth()
-	return 1
```
Lots of copypasta removal since we don't need two version of the proc now (one for mob/living/carbon/human, one for /datum/species)

phil235 commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1022,11 +1010,11 @@
 		return 0
 
-	var/armor = H.run_armor_check(affecting, "melee", "<span class='notice'>Your armor has protected your [hit_area].</span>", "<span class='notice'>Your armor has softened a hit to your [hit_area].</span>",I.armour_penetration)
-	armor = min(90,armor) //cap damage reduction at 90%
```
changed var name because there's already a /datum/species/var/armor

phil235 commented on code/modules/mob/living/carbon/human/update_icons.dm:
```diff
@@ -53,30 +53,10 @@ Please contact me on #coderbus IRC. ~Carnie x
-			if(thing)	overlays += thing
-
-	update_transform()
-
```
Removed all remaining instances of the proc.

MrPerson replied:
Thank you

phil235 commented on code/modules/mob/transform_procs.dm:
```diff
@@ -156,27 +154,21 @@
+	if(tr_flags & TR_KEEPSE)
+		O.dna.struc_enzymes = dna.struc_enzymes
+		var/datum/mutation/human/race/R = mutations_list[RACEMUT] //we don't want to keep the race block active
+		O.dna.struc_enzymes = setblock(O.dna.struc_enzymes, R.dna_block, random_string(DNA_BLOCK_SIZE, list("0","1","2","3","4","5","6")))
```
This ugly line is necessary to avoid having a human with the SE race block above 800, which happens when humanize() is called w/o updating the monkey's SE race block beforehand (changeling's lesser form, admin buttons)

Incoming5643 commented:
If we greyscaled humans and set realistic skin tone bounds we could easily have hulk apply to any greyscaled races just by forcing their color to green and eyes to red. See also lizard bits turning grey to match up with the husked sprite.

Of course then it'd be hard to tell if a lizard was a hulkster if they just happened to be green to begin with.

Lumipharon commented:
Lizard hulks are super bullshit precisely because you can't readily tell that they're hulk.

Does this pr fix the fact that cloning (even a fully upgraded one) doesn't actually clone your dna perfectly?
ie: 
Get scanned in an autocloner
Dna lock a mech
Die
Get cloned
Can't get in mech

PKPenguin321 commented:
PURGE THE LIZARDS
why are lizards compatible with human genes in the first place
:+1:

xxalpha commented:
:-1:

Don't remove the ability for lizards to get hulk.

AnonymousNow commented:
Two things...

- One, it kinda sucks that lizardmen can't be hulks anymore, though I could understand why. Perhaps make their hulk block do something different but still powerful? Make them Dragons or something?
- And two, HEALING HANDS 2015

Supermichael777 commented:
lizard hulk block should totally make them able to hock up fireballs.

Dorsisdwarf commented:
:+1: 
When you play the lizards, you sign up to be an objectively worse character.

phil235 commented:
@Lumipharon 

>- Fixes cloning not setting up correctly dna UE and dna.real_name

See https://github.com/tgstation/-tg-station/pull/11883/files#diff-1e264196f81ead51bb747ac7d38f1a09R188

phil235 commented:
Oh, I almost forgot, monkeys now have blood types and are compatible with humans (mostly because having two separate kinds of blood is more difficult to do). Updated the OP.

SconesC commented:
send the lizard back

also greyscaled humans are probably out of the scope of this pr but nevertheless something that should be looked into

Razharas commented:
Do not rename mech dna locks, just remove them
Mech dna locks are same thing as exosuits special access check - made by coder maining roboticist
REMOVE

Razharas commented on code/datums/mutations.dm:
```diff
@@ -583,32 +591,47 @@
-/mob/living/carbon/human/proc/update_mutation_overlays()
-	remove_overlay(MUTATIONS_LAYER)
-	var/image/standing
+/mob/living/carbon/human/update_mutations_overlay()
```
Im not sure i like this "rewrite all mutations every time you get new one or lose old one"
If you have a list with all the images stored already why not to just substract that list from the overlays_standing[neededlayer] list and then update overlays?

phil235 replied:
will fix.

phil235 commented:
>Do not rename mech dna locks, just remove them
>Mech dna locks are same thing as exosuits special access check - made by coder maining roboticist
>REMOVE

This is work for another PR.

phil235 committed:
Simplified update_mutations_overlay(), we no longer rebuild the entire mutation layer with all mutations every time we get or lose a mutation. I reintroduced in on_acquiring and on_losing some code similar to the old lose_indication gain_indication procs.

Fixed lesser form changeling, when humanized via dna injector, the ling no longer keeps its Human Form power.

Simplified the racemut trick in humanize() and monkeyize().

Dna now gets qdel'd, just like reagents, when the mob gets qdel'd.

Some tiny fixes here and there.

Added a changelog.

phil235 committed:
Tiny mistake, transforming changeling need the target's name.

phil235 commented:
The PR should be ready now.

Razharas commented:
Id prefer the exact overlay handling in the separate proc called in on aq and on lose but its ok
